eip-4626.md

```diff
@@ -0,0 +1,152 @@
+type: Standards Track
+category: ERC
+created: 2021-12-22
+---
+
+## Abstract
+
+The following standard allows for the implementation of a standard API for yield bearing vaults within smart contracts. This standard provides basic functionality for depositing and withdrawing tokens and reading balances with an optional extension for tokenized vaults using ERC-20.
+
+## Motivation
+
+Yield bearing vaults have a lack of standardization leading to diverse implementation details. Some various examples include lending markets (Compound, Aave, Fuse), aggregators (Yearn, Rari Vaults, Idle), and intrinsically interest bearing tokens (xSushi). This makes integration difficult at the aggregator or plugin layer for protocols which need to conform to many standards. This forces each protocol to implement their own adapters which are error prone and waste development resources.
+
+A standard for yield bearing vaults will allow for a similar cambrian explosion to ERC-20, unlocking access to yield in a variety of applications with little specialized effort from developers.
+
+
+## Specification
+
+### Methods
+
+#### deposit
+
+`function deposit(address _to, uint256 _value) public returns (uint256 _shares)`
+
+Deposits `_value` tokens into the vault and grants ownership of them to `_to`.
+
+MAY return a pro-rata ownership `_shares` value corresponding `_value`, if not MUST return `0`.
+
+MAY represent `_shares` using internal accounting or an ERC-20 token. 
+
+If pro-rata shares ownership is implemented, the vault SHOULD implement `balanceOf`, `redeem`, `totalSupply` and `exchangeRate`.
+
+
+#### withdraw
+
+`function withdraw(address _to, uint256 _value) public returns (uint256 _shares)`
+
+Withdraws `_value` tokens from the vault and transfers them to `_to`. 
+
+MAY return a pro-rata ownership `_shares` value corresponding to `_value`,  if not MUST return `0`.
+
+
+#### totalHoldings
+`function totalHoldings() public view returns (uint256)`
+
+Returns the total amount of underlying tokens held/managed by the vault.
+
+
+#### balanceOfUnderlying
+`function balanceOfUnderlying(address _owner) public view returns (uint256)`
+
+Returns the total amount underlying tokens held in the vault for `_owner`.
+
+#### underlying
+`function underlying() public view returns (address)`
+
+Returns the address of the token the vault uses for accounting, depositing, and withdrawing. 
+
+SHOULD return a token implementing the ERC-20 standard.
+
+
+#### totalSupply
+
+`function totalSupply() public view returns (uint256)`
+
+Returns the total number of unredeemed vault shares in circulation.
+
+OPTIONAL - This method is only needed for vaults that implement a pro-rata share mechanism for deposits.
+
+#### balanceOf
+
+`function balanceOf(address _owner) public view returns (uint256)`
+
+Returns the total amount of vault shares the `_owner` currently has.
+
+OPTIONAL - This method is only needed for vaults that implement a pro-rata share mechanism for deposits.
+
+#### redeem
```

Comment:
Given compound's `mint`/`redeemUnderlying` methods have been replaced by `deposit`/`withdraw` in this EIP, it might make sense to retain parallel structure, and replace `redeem` with `withdrawShares`.

I think this makes the method's use case a bit more clear at a glance, though `redeem` might have a nice ring to it... ¯\_(ツ)_/¯

Reply:
but this is not `redeemUnderlying` it's `redeem`, which has similar behavior even in compound:

<img width="884" alt="Screen Shot 2022-01-04 at 9 28 55 PM" src="https://user-images.githubusercontent.com/26209401/148165196-7fc74786-719b-4d57-bb66-86f5dda39f91.png">

Reply:
Yup its a 1-1 of `redeem` from compound! 

But if there is no `redeemUnderlying`, only `withdraw` and `withdraw`+`redeem` do the same thing, just one is scaled for shares, it might make sense to associate the two. (in the same way compound associates `redeem` and `redeemUnderlying`)

Again, I think `redeem` sounds nicer, `withdrawShares` describes what this does in context of other methods. Very minor hairs being split.

Reply:
fair points

> Very minor hairs being split.

thats t11s for ya ;P